### ISSUE TITLE
iml and Deep Learning with Keras

### ISSUE THREAD

**Issue by viniciusbrei** [christophM/iml#71]
Hi Christoph,

I am sorry again, but the solution you suggest is still not working. I tried to include the Predictor, and also study your (great) tutorial https://hub.mybinder.org/user/christophm-iml-xsr7fw04/notebooks/notebooks/tutorial-intro.ipynb and replicate the instructions, adapting for a Keras DL model. This is the code I tried:

X <- df[which(names(df) != "chg")] # "chg" is my classification variable (0 or 1)
predictor <- Predictor$new(model_keras, data = X, y = df$chg,
predict.fun=function(object, newdata){
newData_x = xgb.DMatrix(data.matrix(newdata), missing = NA)
results<-predict(model, newData_x)
return(results)
}
)
imp = FeatureImp$new(predictor, loss = 'ce')

Error message: Error in predict(model, newData_x) : object 'model' not found

Doubts:

1. Is it correct to use "model_keras" (my Keras model object) in Predictor$new(model_keras... ?
2. With which object from Keras do I replace the "xgb.DMatrix" in "newData_x = xgb.DMatrix(data.matrix(newdata), missing = NA)"?
3. Should I replace "newdata" and "newData_x" with my own data or should I leave the function the way it is?
4. Should I replace "model" in "results<-predict(model, newData_x)" with my Keras object?

Thank you again!

**Comment by christophM** [comment]
In this predict function you are naming the parameter `object` but in the body of the function you use `model`. 

So this fixes this issue: 

```r
X <- df[which(names(df) != "chg")] # "chg" is my classification variable (0 or 1)
predictor <- Predictor$new(model_keras, data = X, y = df$chg,
predict.fun=function(model, newdata){
newData_x = xgb.DMatrix(data.matrix(newdata), missing = NA)
results<-predict(model, newData_x)
return(results)
}
)
imp = FeatureImp$new(predictor, loss = 'ce')
``` 

Btw. please don't open new issues for the same problem but use the same thread.

**Comment by viniciusbrei** [comment]
Christoph,
Thank you for your message. The code above generated the following output:

"Interpretation method:  FeatureImp 
error function: ce

Analysed predictor: 
Prediction task: unknown 


Analysed data:
Sampling from data.frame with 144 rows and 41 columns.

Head of results:
      feature importance.05 importance importance.95 permutation.error
1       age_z             1          1             1                 1
2      educ_z             1          1             1                 1
3 income.t1_z             1          1             1                 1
4 income.t2_z             1          1             1                 1
5  disc_inc_z             1          1             1                 1
6       contr             1          1             1                 1"

As you can see, all features importance were the same: 1. 

Any guess why this is happening? 

Thank you again!

**Comment by christophM** [comment]
You look at classification error (ce). It might be that your model only predicts the same class, which can happen when the data is imbalanced.
You should look at the predictions of your model.

**Comment by viniciusbrei** [comment]
Thank you!
